File: sdk/es/src/stepauth/envelope.ts
import {
  EnvelopeFormatError,
  InvalidSignatureError,
  signComponent,
  validateKeyset,
  verifyComponent,
  type Keyset,
  type SigningKeyset,
} from "./crypto.js";

/**
 * A keyset signature: the signing keyset's id plus one signature per component
 * key, ordered by keyset index.
 */
export interface Signature {
  keysetId: string;
  signatures: string[];
}

/**
 * The signed wrapper every protocol message travels in. `payload` is base64 of
 * the raw message bytes, and signatures cover those exact bytes.
 */
export interface Envelope {
  payload: string;
  signature: Signature;
}

// Decode base64 the way Go's base64.StdEncoding.DecodeString does: skip CR/LF
// anywhere, require the standard alphabet plus correct padding, but tolerate
// non-canonical unused trailing bits ("AB==") the way Go does. A bare
// Buffer.from is far laxer (accepts base64url and unpadded input); a naive
// re-encode-and-compare is far stricter (rejects line wraps and "AB==").
function stdBase64(s: unknown): Uint8Array | undefined {
  if (typeof s !== "string") return undefined;
  const compact = s.replace(/[\r\n]/g, "");
  if (compact.length % 4 !== 0 || !/^[A-Za-z0-9+/]*={0,2}$/.test(compact)) {
    return undefined;
  }
  return new Uint8Array(Buffer.from(compact, "base64"));
}

function strictBase64(s: unknown): Uint8Array {
  const bytes = stdBase64(s);
  if (bytes === undefined) throw new InvalidSignatureError();
  return bytes;
}

/**
 * Sign the payload with every key in the keyset and return the marshaled
 * envelope. Signatures cover the payload bytes exactly as they will be
 * transmitted; nothing between here and the wire may reserialize them.
 */
export function signEnvelope(ks: SigningKeyset, payload: Uint8Array): string {
  if (!ks || ks.keys.length === 0) throw new Error("stepauth: signing keyset is empty");
  if (!ks.keysetId) throw new Error("stepauth: signing keyset has no keysetId");
  validateKeyset(ks);

  const envelope: Envelope = {
    payload: Buffer.from(payload).toString("base64"),
    signature: {
      keysetId: ks.keysetId,
      signatures: ks.keys.map((key) =>
        Buffer.from(signComponent(key, payload)).toString("base64"),
      ),
    },
  };
  return JSON.stringify(envelope);
}

/**
 * Resolve the envelope's keysetId against the signer's active keysets and
 * verify all-of, returning the decoded payload. Every failure throws
 * InvalidSignatureError.
 */
export function verifyEnvelope(envelopeJSON: string, keysets: Keyset[]): Uint8Array {
  let env: Envelope;
  try {
    env = JSON.parse(envelopeJSON) as Envelope;
  } catch {
    throw new InvalidSignatureError();
  }
  if (typeof env?.payload !== "string") throw new InvalidSignatureError();
  if (!Array.isArray(env?.signature?.signatures)) throw new InvalidSignatureError();
  if (env.signature.signatures.some((s) => typeof s !== "string")) {
    throw new InvalidSignatureError();
  }

  const keyset = keysets.find((k) => k.keysetId === env.signature.keysetId);
  if (!keyset || !Array.isArray(keyset.keys)) throw new InvalidSignatureError();
  // All-of over a partial keyset verifies against whatever it happens to
  // carry, so a keyset missing a component (or degraded to zero) must reject.
  try {
    validateKeyset(keyset);
  } catch {
    throw new InvalidSignatureError();
  }
  if (env.signature.signatures.length !== keyset.keys.length) {
    throw new InvalidSignatureError();
  }

  const payload = strictBase64(env.payload);

  keyset.keys.forEach((kk, i) => {
    const sig = strictBase64(env.signature.signatures[i]);
    const pub = strictBase64(kk.pub);
    if (!verifyComponent(kk.alg, pub, payload, sig)) {
      throw new InvalidSignatureError();
    }
  });

  return payload;
}

/**
 * Base64-decode a payload WITHOUT verifying the signature. Only for routing a
 * message to the state needed to verify it; the result is untrusted until
 * verifyEnvelope has run.
 */
export function decodeEnvelopePayload(envelopeJSON: string): Uint8Array {
  let env: Envelope;
  try {
    env = JSON.parse(envelopeJSON) as Envelope;
  } catch {
    throw new EnvelopeFormatError();
  }
  const payload = stdBase64(env?.payload);
  if (payload === undefined) throw new EnvelopeFormatError();
  return payload;
}

File: sdk/go/stepauth/envelope.go
package stepauth

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
)

// Signature is a keyset signature: the signing keyset's id plus one signature
// per component key, ordered by keyset index.
type Signature struct {
	KeysetID   string   `json:"keysetId"`
	Signatures []string `json:"signatures"`
}

// Envelope is the signed wrapper every protocol message travels in. Payload is
// base64 of the raw message bytes, and signatures cover those exact bytes.
type Envelope struct {
	Payload   string    `json:"payload"`
	Signature Signature `json:"signature"`
}

// SignEnvelope signs payload with every key in the keyset and returns the
// marshaled envelope. Signatures cover the payload bytes exactly as they will
// be transmitted; nothing between here and the wire may reserialize them.
func SignEnvelope(ks *SigningKeyset, payload []byte) ([]byte, error) {
	if err := ks.validate(); err != nil {
		return nil, err
	}

	sigs := make([]string, len(ks.Keys))
	for i, key := range ks.Keys {
		sig, err := signComponent(key, payload)
		if err != nil {
			return nil, fmt.Errorf("stepauth: signing with key %d (%s): %w", key.Idx, key.Alg, err)
		}
		sigs[i] = base64.StdEncoding.EncodeToString(sig)
	}

	return json.Marshal(Envelope{
		Payload:   base64.StdEncoding.EncodeToString(payload),
		Signature: Signature{KeysetID: ks.KeysetID, Signatures: sigs},
	})
}

// VerifyEnvelope resolves the envelope's keysetId against the signer's active
// keysets and verifies all-of, returning the decoded payload. Every failure
// yields ErrInvalidSignature.
func VerifyEnvelope(envelopeJSON []byte, keysets []Keyset) ([]byte, error) {
	var env Envelope
	if err := json.Unmarshal(envelopeJSON, &env); err != nil {
		return nil, fmt.Errorf("%w: envelope is not valid json", ErrInvalidSignature)
	}

	keyset := keysetByID(keysets, env.Signature.KeysetID)
	if keyset == nil {
		return nil, ErrInvalidSignature
	}
	// Positive check: the keyset must carry exactly the v1 algorithms in
	// order. That subsumes both a zero-key keyset (fails the length check)
	// and a duplicate-algorithm keyset (fails the order check).
	if err := validateKeyset(*keyset); err != nil {
		return nil, ErrInvalidSignature
	}
	if len(env.Signature.Signatures) != len(keyset.Keys) {
		return nil, ErrInvalidSignature
	}

	payload, err := base64.StdEncoding.DecodeString(env.Payload)
	if err != nil {
		return nil, ErrInvalidSignature
	}

	for i, kk := range keyset.Keys {
		sig, err := base64.StdEncoding.DecodeString(env.Signature.Signatures[i])
		if err != nil {
			return nil, ErrInvalidSignature
		}
		pub, err := base64.StdEncoding.DecodeString(kk.Pub)
		if err != nil {
			return nil, ErrInvalidSignature
		}
		if err := verifyComponent(kk.Alg, pub, payload, sig); err != nil {
			return nil, ErrInvalidSignature
		}
	}
	return payload, nil
}

// DecodeEnvelopePayload base64-decodes a payload WITHOUT verifying the
// signature. Only for routing a message to the state needed to verify it; the
// result is untrusted until VerifyEnvelope has run.
func DecodeEnvelopePayload(envelopeJSON []byte) ([]byte, error) {
	var env Envelope
	if err := json.Unmarshal(envelopeJSON, &env); err != nil {
		return nil, fmt.Errorf("stepauth: parsing envelope: %w", err)
	}
	payload, err := base64.StdEncoding.DecodeString(env.Payload)
	if err != nil {
		return nil, fmt.Errorf("stepauth: envelope payload is not valid base64: %w", err)
	}
	return payload, nil
}

func keysetByID(keysets []Keyset, id string) *Keyset {
	for i := range keysets {
		if keysets[i].KeysetID == id {
			return &keysets[i]
		}
	}
	return nil
}
